core/src/main/scala/ai/h2o/sparkling/backend/converters/DataTypeConverter.scala

```diff
@@ -21,37 +21,69 @@ import ai.h2o.sparkling.backend.utils.SupportedTypes
 import ai.h2o.sparkling.extensions.serde.ExpectedTypes
 import ai.h2o.sparkling.extensions.serde.ExpectedTypes.ExpectedType
 import org.apache.spark.ExposeUtils
-import org.apache.spark.rdd.RDD
-import org.apache.spark.sql.Row
+import org.apache.spark.sql.{DataFrame, Row}
 import org.apache.spark.sql.types._
+import org.apache.spark.sql.functions._
 import water.fvec.Vec
-import water.parser.{BufferedString, PreviewParseWriter}
+import water.parser.{BufferedString, Categorical, PreviewParseWriter}
 
 private[backend] object DataTypeConverter {
 
-  private def stringTypesToExpectedTypes(rdd: RDD[Row], schema: StructType): Map[Int, ExpectedType] = {
+  private def stringTypesToExpectedTypes(dataFrame: DataFrame): Map[Int, ExpectedType] = {
     val stringTypeIndices = for {
-      (field, index) <- schema.fields.zipWithIndex
+      (field, index) <- dataFrame.schema.fields.zipWithIndex
       if field.dataType == StringType
     } yield index
 
+    val rdd = dataFrame.rdd
     val types = if (rdd.getNumPartitions > 0) {
       val serializedPreview = rdd
         .mapPartitions[Array[Byte]](createPartitionPreview(_, stringTypeIndices))
         .reduce(mergePartitionPreview)
 
       val preview = CategoricalPreviewWriter.deserialize(serializedPreview)
-      preview.guessTypes().map {
+      val typesFromPreview = preview.guessTypes().map {
         case Vec.T_CAT => ExpectedTypes.Categorical
         case _ => ExpectedTypes.String
       }
+      correctCategoricalTypesForTooHighCardinality(dataFrame, typesFromPreview)
     } else {
       stringTypeIndices.map(_ => ExpectedTypes.String)
     }
 
     stringTypeIndices.zip(types).toMap
   }
 
+  private def correctCategoricalTypesForTooHighCardinality(
+      dataFrame: DataFrame,
+      expectedTypes: Array[ExpectedType]): Array[ExpectedType] = {
+    val stringColumns = dataFrame.schema.fields.filter(_.dataType == StringType).map(_.name).zipWithIndex
+    val categoricalColumns = stringColumns.zip(expectedTypes).filter(_._2 == ExpectedTypes.Categorical).map(_._1)
+
+    if (categoricalColumns.nonEmpty) {
+      val (categoricalNames, categoricalIndices) = categoricalColumns.unzip
+
+      val allowedError = 0.05 // Spark default on approx_count_distinct
+
+      // Adding margin 0.01 just to be sure that the real distinct count won't be under Categorical.MAX_CATEGORICAL_COUNT
+      // and the approximate count won't be above threshold.
+      val threshold = Categorical.MAX_CATEGORICAL_COUNT * (1 + allowedError + 0.01)
```

Comment:
since the count is approximate should't you lower the threshold to make sure the count fits into the threshold?

Reply:
IMHO, the threshold should be higher than `Categorical.MAX_CATEGORICAL_COUNT` if we want to get a similar bahavior as in H2O.  If the true distinct count is exactly `Categorical.MAX_CATEGORICAL_COUNT` the function can return `approx_count_distinct` can return a higher number. With a lower threshold, we would lose some categorical columns. On the other hand, categorical columns with cardinality close to  `Categorical.MAX_CATEGORICAL_COUNT` might not to make sense anaway.

Or did you mean it from the point of view of avoiding the exception? (long strings and high cardinality)?

Reply:
my point was making 100% sure to avoid breking the limit, the limit itself is pretty much theretical and I don't think you want to come close it